File: src/algorithmic_toolbox/week3/CarFueling.java
package algorithmic_toolbox.week3;

import java.util.Scanner;

public class CarFueling {
  
  public static int floor(int[] stops, int val) {
    int start = 0;
    int end = stops.length - 1;

    int mid;
    while (start <= end) {
      mid = (start + end) / 2;
      if (stops[mid] < val)      { start = mid + 1; }
      else if (stops[mid] > val) { end = mid - 1; }
      else                       { return stops[mid]; }
    }

    mid = (start + end) / 2;
    return (stops[mid] < val) ? stops[mid] : stops[mid - 1];
  }

  public static int count_stops(int[] stops, int range) {
    int pos = 0;
    int stopsCounter = 0;

    int reach, newPos;
    while (true) {
      reach = pos + range;
      newPos = floor(stops, reach);
      
      if (newPos == pos) {
        return -1;  // impossible
      }
      pos = newPos;

      if (pos >= stops[stops.length - 1]) {
        break;
      }
      stopsCounter++;
    }

    return stopsCounter;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int dest = scanner.nextInt();
    int range = scanner.nextInt();
    int stopsCount = scanner.nextInt();
    
    int[] stops = new int[stopsCount + 2];
    for (int i = 1; i < stopsCount + 1; i++) {
      stops[i] = scanner.nextInt();
    }
    scanner.close();

    // prepare the 'stops' array
    stops[0] = 0;  // initial pos
    stops[stopsCount + 1] = dest;  // final destination

    System.out.println(count_stops(stops, range));
  }
}

File: src/algorithmic_toolbox/week3/car_fueling.py
from typing import List, TypeVar

T = TypeVar('T')
def floor(lst: List[T], val: T) -> T:
    start = 0
    end = len(lst) - 1

    while start <= end:
        mid = (start  + end) // 2

        if lst[mid] < val:
            start = mid + 1
        elif lst[mid] > val:
            end = mid - 1
        else:
            return lst[mid]
    
    mid = (start + end) // 2
    return lst[mid] if lst[mid] < val else lst[mid - 1]

def calc_refuel(max_range: int, stations: List[int]):
    pos = 0
    stops = 0

    while True:
        reach = pos + max_range
        new_pos = floor(stations, reach)

        if pos == new_pos:
            return -1  # impossible
        pos = new_pos
        
        if new_pos >= stations[-1]:
            break
        stops += 1

    return stops


if __name__ == '__main__':
    dest = int(input())
    max_range = int(input())
    input()
    stations: List[int] = [int(x) for x in input().split()]  # sorted

    stations.insert(0, 0)  # start at position 0
    stations.append(dest)

    print(calc_refuel(max_range, stations))
